### ISSUE TITLE
Support drivers that support both ephemeral and non ephemeral modes

### ISSUE THREAD

**Issue by kfox1111** [kubernetes/kubernetes#79624]
We would like to have drivers that can be used in both ephemeral and non ephemeral modes be supported.

csi-driver-hostpath is an example that can be used in either ephemeral or non ephemeral modes, but only one at a time.

To support both at one time, the driver needs to be aware of which mode its operating on, on a per volume basis. One suggestion would be to use the existing pod info on mount feature (20190129-csi-pod-info-on-mount.md) to pass the property:
csi.storage.k8s.io/ephemeral=true when the volumeHande was provisioned ephemerally and csi.storage.k8s.io/ephemeral=false when normally provisioned.

The KEP 20190122-csi-inline-volumes.md also has a flag specified on the CSIDriver.Mode supporting persistent or ephemeral. A third option could be added to allow Kubernetes to know that it supports both modes of operation at the same time.

**Comment by kfox1111** [comment]
@kubernetes-sig-storage-feature-requests

**Comment by kfox1111** [comment]
@msau42 @vladimirvivien @pohly

**Comment by kfox1111** [comment]
@kubernetes/sig-storage-feature-requests

**Comment by pohly** [comment]
How does kubelet decide whether it shall add that `csi.storage.k8s.io/ephemeral` key to the `NodePublishVolumeRequest.volume_context`? I think it has to depend on that (to be implemented) `CSIDriver.Mode`, because otherwise we end up modifying the `volume_context` also for drivers which don't support ephemeral, which may surprise them.

**Comment by kfox1111** [comment]
The csi pod info on mount spec covers that. The pod info is passed to the driver only if the CSIDriver (https://kubernetes-csi.github.io/docs/csi-driver-object.html)'s podInfoOnMount flag is set to true. maybe we just keep that convention. If the driver doesnt care to know, then it dosn't set podInfoOnMount.

So, I think it would be something like:
```yaml
apiVersion: storage.k8s.io/v1beta2
kind: CSIDriver
metadata:
  name: mycsidriver.example.com
spec:
  attachRequired: false
  podInfoOnMount: true
  mode: ephemeral+persistent
```

If podInfoOnMount is true, we always pass csi.storage.k8s.io/ephemeral. No reason I think to restrict it to only if mode is certain types.

The other way is to force override podInfoOnMount to always be true if mode: ephemeral+persistent. That may be a little surprising though.

**Comment by pohly** [comment]
> If podInfoOnMount is true, we always pass csi.storage.k8s.io/ephemeral. No reason I think to restrict it to only if mode is certain types.

So this basically extends what's considered "pod info". Works for me.

**Comment by pohly** [comment]
Let's update the KEP to mirror the discussion here: https://github.com/kubernetes/enhancements/pull/1129

**Comment by saad-ali** [comment]
> So, I think it would be something like:
> ```yaml
> apiVersion: storage.k8s.io/v1beta2
> kind: CSIDriver
> metadata:
>   name: mycsidriver.example.com
> spec:
>   attachRequired: false
>   podInfoOnMount: true
>   mode: ephemeral+persistent
> ```

This seems weird to me. Would it be possible for you to deploy your driver two times (once in each mode), that way, as far as k8s is concerned, it sees two drivers, one ephemeral and one persistent? That will make it much easier to reason about drivers on the k8s side.

CC @msau42

**Comment by pohly** [comment]
> This seems weird to me. Would it be possible for you to deploy your driver two times (once in each mode), that way, as far as k8s is concerned, it sees two drivers, one ephemeral and one persistent? 

No, that doesn't make sense for a driver that manages some kind of local storage (example: LVM, PMEM). That driver has to keep track of that local storage and hand it out to both persistent and ephemeral volumes. Having to register twice with Kubernetes is unnecessarily complex and also potentially confusing for the user.

**Comment by msau42** [comment]
Can we add that to an "alternatives considered" section in the kep, so we don't lose the history behind the reasoning?

**Comment by kfox1111** [comment]
+1. Could also be added as a finding from the alpha that we changed in the beta to make things better.

**Comment by pohly** [comment]
> Can we add that to an "alternatives considered" section in the kep, so we don't lose the history behind the reasoning?

Done: https://github.com/kubernetes/enhancements/pull/1226
